src/rejson_api.h

```diff
@@ -0,0 +1,71 @@
+#pragma once
+
+#include "redismodule.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef enum JSONType {
+    JSONType_String = 0,
+    JSONType_Int = 1,
+    JSONType_Double = 2,
+    JSONType_Bool = 3,
+    JSONType_Object = 4,
+    JSONType_Array = 5,
+    JSONType_Null = 6,
+    JSONType__EOF
+} JSONType;
+
+typedef const void *RedisJSONKey;
+typedef const void *RedisJSON;
+
+typedef struct RedisJSONAPI_V1 {
+    /* RedisJSONKey functions */
+    RedisJSONKey (*openKey)(RedisModuleCtx *ctx, RedisModuleString *key_name);
+    RedisJSONKey (*openKeyFromStr)(RedisModuleCtx *ctx, const char *path);
+
+    void (*closeKey)(RedisJSONKey key);
+
+    /* RedisJSON functions
+     * Return NULL if path does not exist
+     * `count` can be NULL and return 0 for non array/object
+     **/
+    RedisJSON (*get)(RedisJSONKey key, const char *path, JSONType *type, size_t *count);
```

Comment:
So count represent the number of elements (if the returned object is a collection)?
I think type and count should be 2 distinct functions on 'RedisJSON' object.
I also not sure what the ownership on 'RedisJSON', for how long do I get to keep it?

Reply:
In addition, what if the path returns multiple values?

Reply:
@ashtul I meant that the path might return more than one result, the API should be ready for this. I think that the count should not be here at all, its not relevant to all the types, There should be a separate API for getting the count from the RedisJSON object.

Reply:
@MeirShpilraien `count` is "optional" (could be `NULL`).
1) Currently if the `path` in `get`, or the `index` in `getAt`, lead to an Object or an Array, and if `count` is not `NULL`, the element count at this path/index is returned in this out param. For other types `count` returns zero.
2) For `get` API, a path could yield multiple json values (using wildcards, etc.), but this is not yet supported by RedisJSON (only first result is currently returned).
We can either enhance it to support multi-values although we only need a single now, or wait with that until multi-values are supported, and do it in version 2 of the API.

Reply:
> Currently if the path in get, or the index in getAt, lead to an Object or an Array, and if count is not NULL, the element count at this path/index is returned in this out param. For other types count returns zero.

I believe that count should not be in this API at all, we can introduce a Count function that gets a RedisJSON object. Lets discuss tomorrow on the interactive review.

>For get API, a path could yield multiple json values (using wildcards, etc.), but this is not yet supported by RedisJSON (only first result is currently returned).
We can either enhance it to support multi-values although we only need a single now, or wait with that until multi-values are supported, and do it in version 2 of the API.

We do support finding multiple path, only the user API do not support it and I see no reason why the LLAPI should not support it.